test/jdk/javax/swing/JComponent/bug4247610.java

```diff
@@ -0,0 +1,142 @@
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 4247610
+ * @summary Tests an unnecessary repaint issue
+ * @key headful
+ * @run main bug4247610
+ */
+
+import java.awt.Dimension;
+import java.awt.FlowLayout;
+import java.awt.Graphics;
+import java.awt.Point;
+import java.awt.Robot;
+import java.awt.event.InputEvent;
+import javax.swing.JButton;
+import javax.swing.JDesktopPane;
+import javax.swing.JFrame;
+import javax.swing.JInternalFrame;
+import javax.swing.JLabel;
+import javax.swing.JPanel;
+import javax.swing.SwingUtilities;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.Random;
+
+public class bug4247610 {
+
+    private static JFrame frame;
+    private static JButton damager;
+    private static volatile Point loc;
+    private static volatile Dimension size;
+
+    public static void main(String[] args) throws Exception {
+        Robot robot = new Robot();
+        SwingUtilities.invokeAndWait(() -> {
+            frame = new JFrame("bug4247610");
+            JDesktopPane pane = new JDesktopPane();
+
+            JInternalFrame jif = new JInternalFrame(
+                                 "Damager", true, true, true, true);
+            InternalFramePanel ifp = new InternalFramePanel();
+            damager = new JButton("Damage!");
+            ifp.add(damager);
+            jif.setContentPane(ifp);
+            jif.setBounds(0, 0, 300, 300);
+            jif.setVisible(true);
+            pane.add(jif);
+
+            jif = new JInternalFrame("Damagee", true, true, true, true);
+            JPanel panel = new JPanel(new FlowLayout(FlowLayout.LEFT));
+            final JLabel damagee = new JLabel("");
+            panel.add(damagee);
+            jif.setContentPane(panel);
+            jif.setBounds(60, 220, 300, 100);
+            jif.setVisible(true);
+            pane.add(jif);
+
+            final Random random = new Random();
+
+            damager.addActionListener((e) -> {
+                System.out.println("traceRepaints called");
+                traceRepaints();
+                damagee.setText(Integer.toString(random.nextInt()));
+            });
+            frame.setContentPane(pane);
+            frame.setSize(500, 500);
+            frame.setLocationRelativeTo(null);
+            frame.setVisible(true);
+        });
+        robot.waitForIdle();
+        robot.delay(1000);
+        SwingUtilities.invokeAndWait(() -> {
+            loc = damager.getLocationOnScreen();
+            size = damager.getSize();
+        });
+        robot.mouseMove(loc.x + size.width / 2, loc.y + size.height / 2);
+        robot.waitForIdle();
+        robot.delay(200);
+        robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+        robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
+        if (failed()) {
+            throw new RuntimeException("Failed: unnecessary repaint occured");
+        }
+    }
+
+    private static volatile boolean traceFlag;
+    private static volatile boolean passFlag = true;
```

Comment:
Can be moved to the top along with other variable declarations

Reply:
ok..done..